core/src/worker/workflow/mod.rs

```diff
@@ -1273,3 +1304,22 @@ impl LocalActivityRequestSink for LAReqSink {
         res
     }
 }
+
+/// Sorts jobs in an activation to be in the order lang expects:
+/// `patches -> signals -> other -> queries`
+fn sort_act_jobs(wfa: &mut WorkflowActivation) {
+    wfa.jobs.sort_by(|j1, j2| {
+        // Unwrapping is fine here since we'll never issue empty variants
+        let j1v = j1.variant.as_ref().unwrap();
+        let j2v = j2.variant.as_ref().unwrap();
+        if discriminant(j1v) == discriminant(j2v) {
+            return Ordering::Equal;
+        }
+        match (j1v, j2v) {
+            (workflow_activation_job::Variant::NotifyHasPatch(_), _) => Ordering::Less,
+            (workflow_activation_job::Variant::SignalWorkflow(_), _) => Ordering::Less,
```

Comment:
What if `j1v` is `SignalWorkflow` and `j2v` is `NotifyHasPatch`? I think you should give integer ordinals to each job and then `cmp` them.

Reply:
Yeah, that's a reasonable way to do it. Changed.